src/rusty/src/rest_downloader.rs

```diff
@@ -0,0 +1,325 @@
+                        Err(_) => return None,
+                    }
+                }
+                if chunk_remaining != 0 {
+                    bytes_read -= bytes_skipped;
+                    resp.drain(actual_len..actual_len + bytes_skipped);
+                    if actual_len + chunk_remaining > max_resp { return None; }
+                    let already_in_chunk = cmp::min(bytes_read - actual_len, chunk_remaining);
+                    actual_len += already_in_chunk;
+                    chunk_remaining -= already_in_chunk;
+                    continue 'read_bytes;
+                } else {
+                    if finished_read {
+                        // Note that we may leave some extra \r\ns to be read, but that's OK,
+                        // we'll ignore then when parsing headers for the next request.
+                        resp.resize(actual_len, 0);
+                        return Some(resp);
+                    } else {
+                        // Need to read more bytes to figure out chunk length
+                    }
+                }
+            }
+            resp.resize(bytes_read + cmp::max(10, chunk_remaining), 0);
+            let avail = read_socket!();
+            bytes_read += avail;
+            if chunk_remaining != 0 {
+                let chunk_read = cmp::min(chunk_remaining, avail);
+                chunk_remaining -= chunk_read;
+                actual_len += chunk_read;
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn init_fetch_rest_blocks(uri: *const u8) -> bool {
+    let mut uri_vec = Vec::with_capacity(50);
+    for i in 0..10000 {
+        let c = unsafe { *uri.offset(i) };
+        if c == 0 { break; }
+        uri_vec.push(c);
+    }
+    let uri_str = match String::from_utf8(uri_vec) {
+        Ok(r) => r,
+        Err(_) => return false,
+    };
+    // Sadly only non-SSL is supported for now
+    if let Some((false, _, _, _)) = split_uri(&uri_str) { } else { return false; }
+    std::thread::spawn(move || {
+        // Always catch panics so that even if we have some bug in our parser we don't take the
+        // rest of Bitcoin Core down with us:
+        let _ = catch_unwind(move || {
+            await_ibd_complete_or_stalled();
+            let (ssl, host, port, path) = split_uri(&uri_str).unwrap();
+            let mut provider_state = BlockProviderState::new_with_current_best(BlockIndex::tip());
+            'reconnect: while unsafe { !rusty_ShutdownRequested() } {
+                std::thread::sleep(Duration::from_secs(1));
+                if unsafe { rusty_ShutdownRequested() } { return; }
+
+                let mut stream;
+                macro_rules! reconnect {
+                    () => { {
+                        stream = match TcpStream::connect_timeout(&match (host, port).to_socket_addrs() {
+                            Ok(mut sockaddrs) => match sockaddrs.next() { Some(sockaddr) => sockaddr, None => continue 'reconnect },
```

Comment:
Ideally we would match against the exact error match so that if it returns something like `InvalidData` then there's no need to keep the loop running. (so stop unless it's a timeout, broken pipe, would block etc.)

Reply:
Ehh, I'd rather keep polling. Better to keep polling so that when the rest endpoint fixes their misconfiguration (eg its returning a 500 or 404 or whatever) the downloader picks back up.